### PR TITLE
Spark-Mesos plugin implementation

### PR DESCRIPTION
**Pull request by betobetodev**
This merge request contains the full implementation and changes needed to run an Spark application over Mesos in an infrastructure provided by OpenNebula. There are some points that are very specific to work with the infrastructure security limitations that we have in the context of BigSea EuBra partners. It means that this plugin is not completely generic, because it works following the GryCAP network topology and security policies. But this is a great base to write any other plugin to do similar tasks.

This plugin submit (via ssh) an spark application with the input and output stored in some accessible HDFS. The job binary must be in a public download link or also stored in the HDFS. It also can enable the QoS monitor and the controller services, described in the submission configuration file (JSON).

### CODE REVIEW

**Comment by igornatanael** [comment]
There are some things that I'd be happy to have here before merge it. 

1 - A client example. This point is mandatory;
2 - A README for the plugin in its main directory, with a brief documentation about the plugin. It is valid for the other plugins but must be and specific merge request for each plugin. This one is less essential, but would be a great advance for our documentation.

File: .gitignore [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 .tox/
+manager.cfg
 Application_Manager.egg-info/
 dist/
 build/
 *.pyc
-manager.cfg
 manager.cfg.lsd
-*.un~
+*.un
```
Comment (tellesnobrega) [inline .gitignore]: why change this?

File: application_manager/plugins/spark_mesos/plugin.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                              '--class %s %s %s')
+        else:
+            binary_path = '~/exec_bin.py'
+            spark_run = ('sudo %s --name %s ' +
+                              '--master mesos://%s:%s ' +
+                              '%s %s %s')
+
+        # DONE: Download the binary from a public link
+
+        print ">>>>>>>>>>>>>>>>> ID: " + self.app_id
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove this line and add a proper LOG message

File: application_manager/plugins/spark_mesos/plugin.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        try:
+#           stdin, stdout, stderr = conn.exec_command('wget %s -O %s' %
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove comments here.

File: application_manager/plugins/spark_mesos/plugin.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+#                                                     (binary_url,
+#                                                      binary_path))
+            print "waiting for download the binary"
+#           stdout.read()
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Is it necessary to hold the execution until the end?

File: application_manager/plugins/chronos/plugin.py [hunk 5]
```diff
@@ -40,7 +40,7 @@ def start_application(self, data):
             url = api.chronos_url
```
Comment (tellesnobrega) [inline application_manager/plugins/chronos/plugin.py]: I'm not sure the chronos plugin changes should be merged together with a pull request named Mesos-spark plugin implementation, but I won't block merging on this.

File: application_manager/plugins/spark_mesos/plugin.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+                                                 self.app_id,
+                                                 api.mesos_url,
+                                                 api.mesos_port,
+                                                 execution_class,
+                                                 binary_path,
+                                                 execution_parameters))
+
+        # print 'ERROR RUNNING APPLCATION: ' + e.read()
+
+        # TODO: Discovery ips of the executors from Mesos
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove TODOs

File: application_manager/plugins/spark_mesos/plugin.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+                              (id, api.one_username, api.one_password, api.one_url)
+
+                stdin, stdout, stderr = conn.exec_command(vm_info_one)
+                if ip in stdout.read():
+                    executors_vms_ids.append(id)
+                    break
+
+        print executors_vms_ids
+
+        # TODO: set up the initial configuration of cpu cap
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove TODO e transform in a common comment

File: application_manager/plugins/spark_mesos/plugin.py (hunk 2 above)
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: Do we need this print?

File: application_manager/plugins/spark_mesos/plugin.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        scaler.setup_environment(api.controller_url, executors_vms_ids,
+                                 starting_cap, data)
+
+        # TODO: start monitor service
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove TODO e transform in a common comment

File: application_manager/plugins/spark_mesos/plugin.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+                       "expected_time": expected_time,
+                       "number_of_jobs": number_of_jobs}
+
+        print "%s | Starting monitor" % (time.strftime("%H:%M:%S"))
+        monitor.start_monitor(api.monitor_url, self.app_id,
+                              'spark-mesos', info_plugin, 2)
+
+        # self._start_monitoring(executors[1], data)
+
+        # TODO: start controller service
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove TODO e transform in a common comment

File: application_manager/plugins/spark_mesos/plugin.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+                              '%s %s %s')
+
+        # DONE: Download the binary from a public link
+
+        print ">>>>>>>>>>>>>>>>> ID: " + self.app_id
+        try:
+#           stdin, stdout, stderr = conn.exec_command('wget %s -O %s' %
+#                                                     (binary_url,
+#                                                      binary_path))
+            print "waiting for download the binary"
```
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: shouldn't this be logged?

File: application_manager/plugins/spark_mesos/plugin.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+#           stdout.read()
+            print "binary downloaded"
```
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: ditto

File: application_manager/plugins/spark_mesos/plugin.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+    def to_dict(self):
+        return {
+            'name': self.name,
+            'title': self.get_title(),
+            'description': self.get_description(),
+        }
+
+    def execute(self, data):
+        # mesos_url = api.mesos_url
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Remove these commented lines

File: application_manager/plugins/spark_mesos/plugin.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+#           stdin, stdout, stderr = conn.exec_command('wget %s -O %s' %
+#                                                     (binary_url,
+#                                                      binary_path))
+            print "waiting for download the binary"
+#           stdout.read()
+            print "binary downloaded"
+
+        except Exception as e:
+            print "Error downloading binary"
+            print e.message
```
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: ditto

File: application_manager/plugins/spark_mesos/plugin.py (hunk 7 above)
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: is it still TODO?

File: application_manager/service/api/v10.py [hunk 14]
```diff
@@ -25,11 +25,11 @@
 
 
 def execute(data):
-    authorization = authorizer.get_authorization(api.authorization_url,
-                                                 data['bigsea_username'],
-                                                 data['bigsea_password'])
-    if not authorization['success']:
-        return 'Error: Authentication failed. User not authorized'
+    # authorization = authorizer.get_authorization(api.authorization_url,
```
Comment (igornatanael) [inline application_manager/service/api/v10.py]: Is is working well? If affirmative, remove commented lines.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 8 above)
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: Same question here

File: application_manager/plugins/spark_mesos/plugin.py (hunk 9 above)
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: same

File: application_manager/plugins/spark_mesos/plugin.py [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+                            executors_vms_ids,
+                            data)
+
+        # self._start_controller(executors_vms_ids, data)
+
+        # This command locks the plugin execution until the execution be done
+        print o.read()
+        # TODO: stop monitor
+        # monitor.stop_monitor(api.monitor_url, self.app_id)
+        # TODO: stop controller
```
Comment (tellesnobrega) [inline application_manager/plugins/spark_mesos/plugin.py]: same here

File: setup.py [hunk 16]
```diff
@@ -43,7 +43,8 @@
             'sahara=application_manager.plugins.sahara.plugin:SaharaProvider',
             'fake=application_manager.plugins.fake.plugin:FakeProvider',
             'os_generic=application_manager.plugins.openstack_generic.plugin:OpenStackGenericProvider',
-            'chronos=application_manager.plugins.chronos.plugin:ChronosGenericProvider'
+            'chronos=application_manager.plugins.chronos.plugin:ChronosGenericProvider',
+            'spark_mesos=application_manager.plugins.spark_mesos.plugin:SparkMesosProvider'
```
Comment (igornatanael) [inline setup.py:47]: I think we could change the plugin name and put something more specific like: spark-mesos-upv

File: application_manager/plugins/spark_mesos/plugin.py [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+from application_manager.utils import scaler
+
+from uuid import uuid4
+
+import json
+import paramiko
+import time
+
+
+class SparkMesosProvider(base.PluginInterface):
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Change the name for something more specific as SparkMesosUPV

File: setup.sh [hunk 18]
```diff
@@ -2,5 +2,6 @@ sudo apt-get update
 
 sudo apt-get -y install libffi-dev
 sudo apt-get -y install libssl-dev
-sudo apt-get -y install python-tox
```
Comment (igornatanael) [inline setup.sh]: I think it is an accidental removal

File: application_manager/plugins/spark_mesos/plugin.py [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+        # DONE: look for app-id into the labels and
+        # DONE: get the framework that contains it
+        for t in framework['tasks']:
+            for s in t['statuses']:
+                for n in s['container_status']['network_infos']:
+                    for i in n['ip_addresses']:
+                        # TODO: it must return a list with tuples (ip, host)
+                        executors_ips.append(i['ip_address'])
+
+        print executors_ips, framework['webui_url']
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Transform this print line in a proper LOG message

File: application_manager/plugins/spark_mesos/plugin.py [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+
+        time.sleep(5)
+        stdin, stdout, sterr = conn.exec_command(frameworks_call)
+
+        try:
+            output = stdout.read()
+            mesos_resp = json.loads(output)
+            # print mesos_resp
+        except Exception as e:
+            print e.message
```
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Transform this print line in a proper LOG message

File: application_manager/plugins/spark_mesos/plugin.py (hunk 10 above)
Comment (igornatanael) [inline application_manager/plugins/spark_mesos/plugin.py]: Transform this print line in a proper LOG message

File: application_manager/plugins/chronos/plugin.py (hunk 5 above)
Comment (igornatanael) [reply application_manager/plugins/chronos/plugin.py]: I agree with you. I changed it during the changes for mesos, but I'll move this for another pull request.

File: .gitignore (hunk 1 above)
Comment (betobetodev) [reply .gitignore]: I think this change should be ignored.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 15 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 9 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 8 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 7 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: No. Fixed.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 13 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 11 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 10 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 2 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Fixed.

File: tox.ini [hunk 21]
```diff
@@ -4,11 +4,11 @@
 # and then run "tox" from this directory.
 
 [tox]
-envlist = py27, py35, pep8
+envlist = py27, pep8
 
 [testenv]
 usedevelop = True
-commands = py.test
+#commands = py.test
```
Comment (armstrongmsg) [inline tox.ini]: How about this commented line? Should it be removed?

File: setup.sh [hunk 22]
```diff
@@ -2,5 +2,6 @@ sudo apt-get update
...
+sudo apt-get -y install python-pip
```
Comment (armstrongmsg) [inline setup.sh]: Can you explain this change? What is the problem with installing tox via apt-get?

File: application_manager/plugins/spark_mesos/plugin.py [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+            'description': self.get_description(),
+        }
+
+    def execute(self, data):
+        binary_url = str(data['binary_url'])
+        execution_class = str(data['execution_class'])
+        execution_parameters = str(data['execution_parameters'])
+        expected_time = int(data['expected_time'])
+        number_of_jobs = int(data['number_of_jobs'])
+        starting_cap = 0 if data['starting_cap'] == "" \
```
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Why do we need this check? Shouldn't "starting_cap" be always present in "data"? If not, do you realize it is assigning no resources to the application?

File: application_manager/plugins/spark_mesos/plugin.py [hunk 24]
```diff
@@ -0,0 +1,250 @@
...
+
+        try:
+            stdin, stdout, stderr = conn.exec_command('wget %s -O %s' %
+                                                      (binary_url,
+                                                       binary_path))
+
+            plugin_log.log("%s | Waiting for download the binary..."
+                            % (time.strftime("%H:%M:%S")))
+
+            stdout.read()
```
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Does stdout.read() raise an exception when wget fails?

File: application_manager/plugins/spark_mesos/plugin.py [hunk 25]
```diff
@@ -0,0 +1,250 @@
...
+        if key_path != "" and key_path is not None:
+            keypair = paramiko.RSAKey.from_private_key_file(key_path)
+            conn.connect(hostname=ip, username=username, pkey=keypair)
+        else:
+            conn.connect(hostname=ip, username=username, password=password)
+
+        return conn
+
+    def _get_executors_ip(self, conn):
+        frameworks_call = ('curl http://' + self.get_frameworks_url
```
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Does this call return all applications? Or only the running ones?

File: application_manager/plugins/spark_mesos/plugin.py [hunk 26]
```diff
@@ -0,0 +1,250 @@
...
+        # keys or passwords
+        if key_path != "" and key_path is not None:
+            keypair = paramiko.RSAKey.from_private_key_file(key_path)
+            conn.connect(hostname=ip, username=username, pkey=keypair)
+        else:
+            conn.connect(hostname=ip, username=username, password=password)
+
+        return conn
+
+    def _get_executors_ip(self, conn):
```
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: I think this method should be moved to "utils/mesos.py"

File: application_manager/plugins/spark_mesos/plugin.py [hunk 27]
```diff
@@ -0,0 +1,250 @@
...
+
+        plugin_log.log("%s | Executors IPs: %s"
+                        % (time.strftime("%H:%M:%S"), executors_ips))
+
+        plugin_log.log("%s | WebUI URL: %s"
+                        % (time.strftime("%H:%M:%S"), framework['webui_url']))
+
+        return executors_ips, framework['webui_url']
+
+    def _extract_vms_ids(self, output):
```
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: I think this method should be moved to "utils/mesos.py"

File: application_manager/plugins/spark_mesos/plugin.py [hunk 28]
```diff
@@ -0,0 +1,250 @@
...
+        monitor.stop_monitor(api.monitor_url, self.app_id)
+
+        plugin_log.log("%s | Stopping scaler" % (time.strftime("%H:%M:%S")))
+        scaler.stop_scaler(api.controller_url, self.app_id)
+
+        plugin_log.log("%s | Remove binaries" % (time.strftime("%H:%M:%S")))
+        conn.exec_command('rm -rf ~/exec_bin.*')
+        return None, None
+
+    def _get_ssh_connection(self, ip, username=None,
```
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Maybe move this to other file, like "utils/ssh_utils.py"

File: application_manager/plugins/spark_mesos/plugin.py [hunk 29]
```diff
@@ -0,0 +1,250 @@
...
+
+        plugin_log.log("%s | Stopping monitor" % (time.strftime("%H:%M:%S")))
+        monitor.stop_monitor(api.monitor_url, self.app_id)
+
+        plugin_log.log("%s | Stopping scaler" % (time.strftime("%H:%M:%S")))
+        scaler.stop_scaler(api.controller_url, self.app_id)
+
+        plugin_log.log("%s | Remove binaries" % (time.strftime("%H:%M:%S")))
+        conn.exec_command('rm -rf ~/exec_bin.*')
+        return None, None
```
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Why are you returning a tuple of None objects? What is the caller expecting as result?

File: application_manager/plugins/spark_mesos/plugin.py [hunk 30]
```diff
@@ -0,0 +1,250 @@
...
+        return 'It runs an spark application on a Mesos cluster'
+
+    def to_dict(self):
+        return {
+            'name': self.name,
+            'title': self.get_title(),
+            'description': self.get_description(),
+        }
+
+    def execute(self, data):
```
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: Is it possible to split this method in submethods like "get_framework", "get_execution_vm_ids", "get_execution_vm_ips", "start_application"?

**Review by armstrongmsg — CHANGES_REQUESTED** [review]

File: tox.ini (hunk 21 above)
Comment (betobetodev) [inline tox.ini]: I agree.

File: setup.sh (hunk 22 above)
Comment (betobetodev) [reply setup.sh]: What do you think we'll only consider the files with relevant changes to the Mesos plugin? There are many files that do not make sense to change. This is an example. @tellesnobrega @armstrongmsg  @iurygregory @igornatanael

File: application_manager/plugins/spark_mesos/plugin.py (hunk 27 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: I agree.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 26 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: I agree.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 28 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: I agree.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 29 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: The return of 'execute' method of a plugin provider should be the application id and the object of executor.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 30 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Yes. This will be done.

File: .gitignore (hunk 1 above)
Comment (betobetodev) [reply .gitignore]: I search about the ".un" extension and the right form is ".un~". It's a file created by vim.

File: setup.sh (hunk 22 above)
Comment (tellesnobrega) [reply setup.sh]: It makes perfect sense to only change things related to mesos on this pull request

File: application_manager/plugins/spark_mesos/plugin.py (hunk 25 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: I perceive now that this is a semantic problem in Spark-Mesos plugin. To fix without many changes, my suggestion is a 'busy' flag. What do you think?

File: application_manager/plugins/spark_mesos/plugin.py (hunk 23 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: To simplify, consider that when 'starting_cap' doesn't exists in config file, this is set to 100.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 24 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: It's an unusual error. Anyway, this is **TODO: Fix possible wget error**

File: tox.ini (hunk 21 above)
Comment (armstrongmsg) [reply tox.ini]: ?

File: application_manager/plugins/spark_mesos/plugin.py [hunk 31]
```diff
@@ -0,0 +1,260 @@
...
+            plugin_log.log("%s | Starting application execution" %
+                          (time.strftime("%H:%M:%S")))
+
+            binary_url = str(data['binary_url'])
+            execution_class = str(data['execution_class'])
+            execution_parameters = str(data['execution_parameters'])
+            expected_time = int(data['expected_time'])
+            number_of_jobs = int(data['number_of_jobs'])
+            actuator = str(data['actuator'])
+            starting_cap = 100 if data['starting_cap'] == "" \
```
Comment (armstrongmsg) [inline application_manager/plugins/spark_mesos/plugin.py]: I don't think this check makes any sense. You should remove it.

**Review by armstrongmsg — DISMISSED** [review]

File: application_manager/plugins/spark_mesos/plugin.py [hunk 32]
```diff
@@ -0,0 +1,260 @@
+# Copyright (c) 2017 UFCG-LSD and UPV-GRyCAP.
```
Comment (iurygregory) [inline application_manager/plugins/spark_mesos/plugin.py:1]: application_manager/plugins/spark_mesos/plugin.py:23:1: F401 'ID_Generator' imported but unused
from application_manager.utils.ids import ID_Generator
^
application_manager/plugins/spark_mesos/plugin.py:28:1: F401 'json' imported but unused
import json
^
application_manager/plugins/spark_mesos/plugin.py:29:1: F401 'paramiko' imported but unused
import paramiko
^
application_manager/plugins/spark_mesos/plugin.py:73:13: F841 local variable 'actuator' is assigned to but never used
            actuator = str(data['actuator'])
            ^
application_manager/plugins/spark_mesos/plugin.py:75:30: E127 continuation line over-indented for visual indent
                             else int(data['starting_cap'])
                             ^
application_manager/plugins/spark_mesos/plugin.py:77:54: W291 trailing whitespace
            plugin_log.log("%s | Submission id: %s" %
                                                     ^
application_manager/plugins/spark_mesos/plugin.py:84:45: E127 continuation line over-indented for visual indent
                                            api.cluster_username,
                                            ^
application_manager/plugins/spark_mesos/plugin.py:85:45: E127 continuation line over-indented for visual indent
                                            api.cluster_password,
                                            ^
application_manager/plugins/spark_mesos/plugin.py:86:45: E127 continuation line over-indented for visual indent
                                            api.cluster_key_path)
                                            ^
application_manager/plugins/spark_mesos/plugin.py:98:35: E127 continuation line over-indented for visual indent
                                  '--master mesos://%s:%s ' +
                                  ^
application_manager/plugins/spark_mesos/plugin.py:103:35: E127 continuation line over-indented for visual indent
                                  '--master mesos://%s:%s ' +
                                  ^
application_manager/plugins/spark_mesos/plugin.py:113:1: W293 blank line contains whitespace

^
application_manager/plugins/spark_mesos/plugin.py:115:33: E127 continuation line over-indented for visual indent
                                % (time.strftime("%H:%M:%S")))
                                ^
application_manager/plugins/spark_mesos/plugin.py:116:1: W293 blank line contains whitespace

^
application_manager/plugins/spark_mesos/plugin.py:120:33: E127 continuation line over-indented for visual indent
                                % (time.strftime("%H:%M:%S")))
                                ^
application_manager/plugins/spark_mesos/plugin.py:121:1: W293 blank line contains whitespace

^
application_manager/plugins/spark_mesos/plugin.py:124:33: E127 continuation line over-indented for visual indent
                                % (time.strftime("%H:%M:%S")))
                                ^
application_manager/plugins/spark_mesos/plugin.py:139:31: E127 continuation line over-indented for visual indent
                              % (api.one_username,
                              ^
application_manager/plugins/spark_mesos/plugin.py:146:1: W293 blank line contains whitespace

^
application_manager/plugins/spark_mesos/plugin.py:148:60: E127 continuation line over-indented for visual indent
                                                           self.app_id)
                                                           ^
application_manager/plugins/spark_mesos/plugin.py:164:57: E127 continuation line over-indented for visual indent
                                                        api.one_password,
                                                        ^
application_manager/plugins/spark_mesos/plugin.py:165:57: E127 continuation line over-indented for visual indent
                                                        api.one_url)
                                                        ^
application_manager/plugins/spark_mesos/plugin.py:188:80: E501 line too long (80 > 79 characters)
            plugin_log.log("%s | Starting scaler" % (time.strftime("%H:%M:%S")))
                                                                               ^
application_manager/plugins/spark_mesos/plugin.py:206:80: E501 line too long (80 > 79 characters)
            plugin_log.log("%s | Remove binaries" % (time.strftime("%H:%M:%S")))
                                                                               ^
application_manager/plugins/spark_mesos/plugin.py:251:64: E502 the backslash is redundant between brackets
            handling_thread = threading.Thread(target=executor.\
                                                               ^
application_manager/plugins/spark_mesos/plugin.py:259:44: W291 trailing whitespace
        self.running_application = executor


Many pep8 erros

File: application_manager/utils/mesos.py [hunk 33]
```diff
@@ -0,0 +1,66 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+import json
+
+def extract_vms_ids(output):
```
Comment (iurygregory) [inline application_manager/utils/mesos.py]: application_manager/utils/mesos.py:19:1: E302 expected 2 blank lines, found 1
def extract_vms_ids(output):
^

File: application_manager/utils/mesos.py [hunk 34]
```diff
@@ -0,0 +1,66 @@
...
+def get_executors_ip(conn, get_frameworks_url, app_id):
+    frameworks_call = ('curl http://' + get_frameworks_url + '/frameworks')
+
+    time.sleep(5)
+    stdin, stdout, sterr = conn.exec_command(frameworks_call)
+
+    try:
+        output = stdout.read()
+        mesos_resp = json.loads(output)
+    except Exception as e:
```
Comment (iurygregory) [inline application_manager/utils/mesos.py:38]: application_manager/utils/mesos.py:37:25: F841 local variable 'e' is assigned to but never used
    except Exception as e:
                        ^

Talvez atualizaro except para mostrar a mensagem da excecao mesmo

File: application_manager/utils/ssh.py [hunk 35]
```diff
@@ -0,0 +1,33 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paramiko
+
+
+def get_connection(ip, username=None,
+                       password=None, key_path=None):
```
Comment (iurygregory) [inline application_manager/utils/ssh.py]: application_manager/utils/ssh.py:20:24: E127 continuation line over-indented for visual indent
                       password=None, key_path=None):

File: manager.cfg.template [hunk 36]
```diff
@@ -1,31 +1,50 @@
 [services]
-monitor_url = http://0.0.0.0:5001
-controller_url = http://0.0.0.0:5002
-authorization_url = http://0.0.0.0:5003/authorization
-optimizer_url = http://0.0.0.0:5004/optimizer
-plugins = sahara,os_generic
+controller_url =
```
Comment (iurygregory) [inline manager.cfg.template:2]: Não seria melhor manter os valores não? ficaria mais fácil para os usuários (talvez)

**Review by iurygregory — CHANGES_REQUESTED** [review]

File: manager.cfg.template (hunk 36 above)
Comment (betobetodev) [reply manager.cfg.template:2]: manager.cfg.example added.

File: application_manager/utils/mesos.py (hunk 34 above)
Comment (betobetodev) [reply application_manager/utils/mesos.py:38]: Print added.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 32 above)
Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py:1]: Pep8 errors in this file was fixed.

File: setup.sh (hunk 22 above)
Comment (betobetodev) [reply setup.sh]: Done.

File: tox.ini (hunk 21 above)
Comment (betobetodev) [reply tox.ini]: tox.ini was changed to old version.

File: application_manager/plugins/spark_mesos/plugin.py (hunk 31 above)
Comment (robertonscjr) [reply application_manager/plugins/spark_mesos/plugin.py]: Removed.

Comment (betobetodev) [reply application_manager/plugins/spark_mesos/plugin.py]: Removed.

**Review by igornatanael — APPROVED** [review]
It's ok for me.

**Pull request merged**